### ISSUE TITLE
Disable shortcuts

### ISSUE THREAD

**Issue by iqnite** [ScratchAddons/ScratchAddons#6804]
### Why this would be helpful

It's extremely frustrating when you press a key in a project and that key affects the editor.
For example, you surely experienced that when you have a selected object in the paint editor and press the arrow keys in the project, the object suddenly shifts around? Yes, that's because every part of the editor reacts to any key press. 
Another very well-known thing is when the last block you interacted with gets deleted by pressing Alt+X to pause the project in the editor, as "X" is recognized as binding for deleting.

### How this addon works

This addon disables key interactions for areas that are not hovered, as Unity does.

### Possible alternatives

To address the Alt+X issue and similar things, we could also consider disabling those on-key bindings for the codespace. I don't know anybody who uses them, and they are annoying as you first need to execute the block by clicking on it.

### Additional context

I was working on a game when I noticed a big bug that never occurred before. I looked for half an hour for a patch when I suddenly realized that an if-else block with a lot of nested code had mysteriously disappeared. I had to rebuild it from scratch, and that wasn't fun.
Please don't comment "I use those keybinds!" or stuff like that. Thanks.

**Comment by DNin01** [comment]
Clicking on the stage (to focus it) first might resolve the issue.

**Comment by WorldLanguages** [comment]
> For example, you surely experienced that when you have a selected object in the paint editor and press the arrow keys in the project, the object suddenly shifts around? Yes, that's because every part of the editor reacts to any key press.

Can anyone link to the relevant scratch-gui files that add the event listeners to the page? I believe we're looking for different event listeners for the code area and for the costume editor. Does the sound editor have keyboard shortcuts?

**Comment by DNin01** [comment]
> Can anyone link to the relevant scratch-gui files that add the event listeners to the page?

Looks like it's somewhere around here:

https://github.com/scratchfoundation/scratch-blocks/blob/f116732bbbc6b1e0d95119c6ff8d4e23e27da8c0/core/blockly.js#L257

Its parent function is on [line 185](https://github.com/scratchfoundation/scratch-blocks/blob/f116732bbbc6b1e0d95119c6ff8d4e23e27da8c0/core/blockly.js#L185).

Here's the call stack when deleting a block with the Delete key (from Scratch Desktop 3.27.0):
![Call Stack](https://github.com/ScratchAddons/ScratchAddons/assets/106490990/d515a89e-6e25-45a6-9d3e-0ca7728d72d8)

**Comment by DNin01** [comment]
Oh, I can explain this.

[When a key is pressed](https://github.com/scratchfoundation/scratch-blocks/blob/f116732bbbc6b1e0d95119c6ff8d4e23e27da8c0/core/blockly.js#L185), one of the things Scratch checks for is [if you're holding down <kbd>Alt</kbd>, <kbd>Ctrl</kbd>, or <kbd>Command</kbd>](https://github.com/scratchfoundation/scratch-blocks/blob/f116732bbbc6b1e0d95119c6ff8d4e23e27da8c0/core/blockly.js#L212), then if you are, it checks if you're pressing <kbd>C</kbd>, <kbd>X</kbd>, or <kbd>V</kbd>. <kbd>C</kbd> copies, <kbd>X</kbd> cuts (copies and deletes), and <kbd>V</kbd> pastes.

So what we're actually doing is cutting the block. Typically, this is done with <kbd>Ctrl</kbd>+<kbd>X</kbd> but <kbd>Alt</kbd>+<kbd>X</kbd> also works because that was in the if condition for some reason.

<details><summary>Here's where that happens:</summary>
<p>

```js
Blockly.onKeyDown_ = function(e) {
  // ...
  var deleteBlock = false;
  if (e.keyCode == 27) {
    // ...
  } else if (e.altKey || e.ctrlKey || e.metaKey) {
    // ...
    if (Blockly.selected && Blockly.selected.isDeletable() && Blockly.selected.isMovable()) {
      if (e.keyCode == 67) {
        // 'c' for copy.
        Blockly.hideChaff();
        Blockly.copy_(Blockly.selected);
      } else if (e.keyCode == 88 && !Blockly.selected.workspace.isFlyout) {
        // 'x' for cut, but not in a flyout.
        // Don't even copy the selected item in the flyout.
        Blockly.copy_(Blockly.selected);
        deleteBlock = true;
      }
    }
    if (e.keyCode == 86) {
      // 'v' for paste.
      // ...
    } else if (e.keyCode == 90) {
      // 'z' for undo 'Z' is for redo.
      Blockly.hideChaff();
      Blockly.mainWorkspace.undo(e.shiftKey);
    }
  }
  // Common code for delete and cut.
  // Don't delete in the flyout.
  if (deleteBlock && !Blockly.selected.workspace.isFlyout) {
    Blockly.Events.setGroup(true);
    Blockly.hideChaff();
    Blockly.selected.dispose(/* heal */ true, true);
    Blockly.Events.setGroup(false);
  }
};
```

</p>
</details>

**Comment by iqnite** [comment]
Wait... I thought you could only copy/cut/paste with DevTools?

**Comment by DNin01** [comment]
> Wait... I thought you could only copy/cut/paste with DevTools?

Yeah, Scratch has built-in copy, cut, delete, and paste functionality for the code area. You just need to focus the block you want to copy/cut/delete by clicking it (you can right-click or drag to avoid executing it) and then do the appropriate keyboard shortcut.

**Comment by WorldLanguages** [comment]
I think the devtools addon just presses the shortcut on your behalf.

**Comment by iqnite** [comment]
I've been overestimating its power for years...

> I think the devtools addon just presses the shortcut on your behalf.

**Comment by DNin01** [comment]
From what I've found, it looks like the code area is the only place where <kbd>Alt</kbd>+<kbd>X</kbd> does anything. Everywhere else I've seen, either it's <kbd>Ctrl</kbd>+<kbd>X</kbd> or cutting isn't available, so to resolve the problem of pausing the project cutting a block, that should do it. The only other notable problem I can think of is how arrow keys shift objects in costumes. Anything else to consider? Is that why this issue was made?

**Comment by WorldLanguages** [comment]
There are other related issues that talk about handling keyboard shortcuts/hotkeys and in the future trying to provide a way to customize all vanilla Scratch and addon shortcuts in the same place: #5145 #4946

**Comment by DNin01** [comment]
I'm not an expert on trapping Scratch internals, but would trapping Blockly let us deactivate this feature?

**Comment by WorldLanguages** [comment]
Related feedback, I believe it's a Scratch bug.

> pressing backspace in fullscreen deletes recently interacted block in the editor

**Comment by WorldLanguages** [comment]
This is also somewhat related to this suggestion by griffpatch, regarding the middle-click-popup (insert blocks by name) addon:

![image](https://github.com/ScratchAddons/ScratchAddons/assets/17484114/cd1abc21-b966-4b66-b617-c68ad28af529)

After he suggested this, we tried to discuss how to determine if the user is trying to type into the stage or into the code area.

**Comment by WorldLanguages** [comment]
> I don't know if maybe clicking on the stage (to focus it) first could prevent this from happening? Edit: it does not!

I think that might be intentional, as some users might not notice they need to click the stage...

We should probably try to find a way to make this somewhat intuitive, plus making sure SA users enabling this new addon read the addon description, and it would also have the side-effect of making griffpatch's idea possible to implement.

**Comment by DNin01** [comment]
So, the Blockly `onKeyDown_` function can be accessed like so:
```js
export default async function ({ addon, console }) {
  const Blockly = await addon.tab.traps.getBlockly();
  console.log(Blockly.onKeyDown_);
}
```
But I haven't figured out how to deactivate it yet.

**Comment by WorldLanguages** [comment]
> But I haven't figured out how to deactivate it yet.

Multiple addons do this, basically you replace it with another function that only calls the original function if `addon.self.disabled`

**Comment by DNin01** [comment]
What if you could individually toggle keyboard shortcuts? For example, I could have backspace to delete enabled but turn off all of the Ctrl-key shortcuts which I never use in the Scratch code editor.

**Comment by WorldLanguages** [comment]
> > Off-topic: Sometimes I use Backspace or Delete to delete blocks, but I did not know that x does too! (Apparently, it's actually Alt+x.)
> 
> Ugh, and we thought Alt+X for pausing was a unique combination...

+1 from feedback
I think it needs its own dedicated issue now

**Comment by WorldLanguages** [comment]
I created issue #7225

**Comment by mybearworld** [comment]
[Here](https://github.com/scratchfoundation/scratch-blocks/blob/688d708886b72514b4eedbf5a40df65251b7d78d/core/inject.js#L422), `Blockly.onKeyDown_` is added as an event listener directly. So I don't think changing `Blockly.onKeyDown_` afterwards will do anything.

**Comment by DNin01** [comment]
+1 from feedback
